File: lib/supplier-prices.js
'use strict';

/**
 * What a supplier's prices may be compared against.
 *
 * A purchase in Khayt records an amount, a quantity and a **unit** the shop
 * picked from a list: spool, kg, g, L, piece, roll, box. The material type is
 * free text, so the same word — "PLA" — is attached to a spool bought for 75
 * and, a month later, a kilogram bought for 22.
 *
 * The supplier price history chart put both of those on one trend line,
 * badged the move between them as a price change, and then named a "best
 * price" supplier by sorting the mixed numbers. All three are meaningless. The
 * shop that sells by the gram always wins a comparison against the shop that
 * sells by the spool, and a shop that switches from grams to kilos is told its
 * PLA got a thousand times more expensive.
 *
 * Two prices are comparable when they measure the same thing. That is the
 * whole rule, and it has two halves:
 *
 *   1. Units fall into **families**. `g` and `kg` are both mass, so a price in
 *      one converts into the other exactly. A spool is not a mass — spools
 *      hold different amounts — so `spool` is its own family and converts to
 *      nothing. Same for a piece, a roll, a box and a litre.
 *
 *   2. Within a family, prices convert to the family's **base unit** before
 *      anything is compared, plotted or ranked.
 *
 * So the chart groups by material *and* family, not by material alone. A shop
 * that buys PLA both ways sees two cards, each honest, instead of one that is
 * wrong.
 *
 * Deliberately NOT here: a spool-to-kilogram conversion via an assumed spool
 * weight. A 1kg spool and a 250g sample spool are both "a spool", and guessing
 * which one a purchase meant would put a fabricated number in front of a
 * buying decision. Unconvertible is the correct answer, not a gap to fill.
 *
 * The caller supplies the label for untagged material, because that is a
 * translated string and this module holds no text.
 */
(function (global) {

  /**
   * Unit spellings that mean the same unit.
   *
   * The purchase form offers a fixed list, so in a book written by one machine
   * every unit is already canonical. Books merge across devices and across
   * versions of the app, and a unit typed by an importer or an older build is
   * still a unit, so the aliases are cheap insurance against a stray "Kg"
   * opening a second card for the same material.
   */
  const ALIASES = {
    g: 'g', gram: 'g', grams: 'g', gm: 'g', gs: 'g',
    kg: 'kg', kgs: 'kg', kilo: 'kg', kilos: 'kg', kilogram: 'kg', kilograms: 'kg',
    l: 'L', liter: 'L', liters: 'L', litre: 'L', litres: 'L',
    spool: 'spool', spools: 'spool',
    roll: 'roll', rolls: 'roll',
    piece: 'piece', pieces: 'piece', pcs: 'piece', pc: 'piece',
    box: 'box', boxes: 'box',
  };

  /**
   * How a unit converts within its family.
   *
   * `base` is the unit every price in the family is expressed in before it is
   * compared, and `per` is how many base units one of this unit is. A price is
   * currency *per unit*, so converting it divides rather than multiplies:
   * 0.02 per gram is 0.02 / 0.001 = 20 per kilogram.
   *
   * A unit absent from this table is its own family and converts to itself,
   * which is what makes an unrecognised unit safe: it is never mixed with
   * anything, including another unrecognised one spelled differently.
   */
  const CONVERSIONS = {
    g: { base: 'kg', per: 0.001 },
    kg: { base: 'kg', per: 1 },
  };

  /** The unit a purchase is assumed to be in when it records none. */
  const DEFAULT_UNIT = 'spool';

  /**
   * Fold a recorded unit onto its canonical spelling.
   *
   * An empty or missing unit becomes the default rather than a family of its
   * own, because a book written before the unit field existed holds purchases
   * that were all spools.
   */
  function normalizeUnit(unit) {
    const raw = String(unit == null ? '' : unit).trim();
    if (!raw) return DEFAULT_UNIT;
    const hit = ALIASES[raw.toLowerCase()];
    return hit || raw;
  }

  /** The unit that prices in this unit's family are compared in. */
  function baseUnit(unit) {
    const u = normalizeUnit(unit);
    return (CONVERSIONS[u] || {}).base || u;
  }

  /**
   * Convert a price per `unit` into a price per its family's base unit.
   *
   * Returns the price unchanged for any unit that is its own base, which is
   * every unit except `g`.
   */
  function toBasePrice(price, unit) {
    const u = normalizeUnit(unit);
    const conv = CONVERSIONS[u];
    const p = +price || 0;
    if (!conv || !conv.per) return p;
    return p / conv.per;
  }

  /**
   * What one of something cost.
   *
   * An explicitly recorded unit price wins; otherwise the amount is spread
   * over the quantity. This is the rule the chart used to reach into the
   * renderer's `computeUnitPrice` for; having it here is what lets a host
   * without a renderer, the Mac app, reach the same figure. `computeUnitPrice`
   * stays where it is as an exported helper, and `test/supplier-prices.test.js`
   * holds the two to each other so they cannot drift apart.
   */
  function unitPriceOf(purchase) {
    const p = purchase || {};
    if (p.unitPrice && +p.unitPrice > 0) return +p.unitPrice;
    const qty = +p.quantity || 1;
    return (+p.amount || 0) / qty;
  }

  /**
   * Every purchase across every supplier, grouped into comparable sets.
   *
   * One group per material *and* base unit, sorted by material then unit so
   * the same book always draws the same order. Each group carries:
   *
   *   - `entries`  — every purchase in it, price converted to `unit`, sorted
   *                  oldest first by date. Undated purchases keep their place
   *                  in `all` but are left out of `entries`, because a trend
   *                  line cannot place a point with no date on it.
   *   - `all`      — every purchase in the group including the undated ones,
   *                  which is what `count`, `best` and `worst` are drawn from.
   *   - `pctChange`— the move from the second-newest dated purchase to the
   *                  newest, now necessarily a like-for-like comparison.
   *
   * `opts.untagged` labels purchases with no material type.
   */
  function groups(suppliers, opts) {
    const options = opts || {};
    const untagged = options.untagged || 'Untagged';
    const bins = new Map();

    (suppliers || []).forEach(sup => {
      const supplier = (sup && sup.name) || '';
      ((sup && sup.purchases) || []).forEach(p => {
        const material = String((p && p.materialType) || '').trim() || untagged;
        const unit = normalizeUnit(p && p.unit);
        const base = baseUnit(unit);
        const key = JSON.stringify([material, base]);
        if (!bins.has(key)) bins.set(key, { material, unit: base, all: [] });
        bins.get(key).all.push({
          date: (p && p.date) || '',
          price: toBasePrice(unitPriceOf(p), unit),
          unit: base,
          recordedUnit: unit,
          supplier,
          total: +(p && p.amount) || 0,
        });
      });
    });

    const out = Array.from(bins.values());
    out.sort((a, b) => (a.material.localeCompare(b.material)) || (a.unit.localeCompare(b.unit)));

    return out.map(g => {
      const entries = g.all.filter(e => e.date).slice()
        .sort((a, b) => a.date.localeCompare(b.date));
      const byPrice = g.all.slice().sort((a, b) => a.price - b.price);
      const last = entries[entries.length - 1] || null;
      const prev = entries[entries.length - 2] || null;
      const pctChange = (prev && prev.price)
        ? ((last.price - prev.price) / prev.price) * 100
        : 0;
      return {
        material: g.material,
        unit: g.unit,
        entries,
        all: g.all,
        count: g.all.length,
        latest: last,
        previous: prev,
        pctChange,
        best: byPrice[0] || null,
        worst: byPrice[byPrice.length - 1] || null,
        /** True when the group holds more than one spelling of its base unit. */
        converted: g.all.some(e => e.recordedUnit !== e.unit),
      };
    });
  }

  const api = {
    DEFAULT_UNIT, ALIASES, CONVERSIONS,
    normalizeUnit, baseUnit, toBasePrice, unitPriceOf, groups,
  };
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  global.KhaytSupplierPrices = api;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: mac/KhaytCore/Sources/KhaytApp/SupplierPrices.swift
import SwiftUI
import KhaytCore

/// What the shop has paid for a material, and whether that has moved.
///
/// ── WHY THIS IS NOT ONE TREND LINE PER MATERIAL ───────────────────────────
///
/// The same word — "PLA" — is attached to a spool bought for 75 and, a month
/// later, a kilogram bought for 22. One line through both says the shop's PLA
/// got cheaper when it did nothing of the sort, and a "best price" picked by
/// sorting the mixed numbers always names whoever sells by the smaller unit.
/// `lib/supplier-prices.js` groups by material AND unit family for exactly
/// that reason; this card draws the groups it is given and never merges them.
///
/// A card per group rather than a chart: what a shop does with this is decide
/// who to ring, and the two facts that settle it are what the last one cost
/// and who has been cheapest.
struct SupplierPricesCard: View {
    @Bindable var shop: Shop
    let groups: [KhaytEngine.PriceGroup]

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            CapsLabel(shop.words.callIt("mac.price_history"), tint: Role.text3, size: 9)
            VStack(spacing: 0) {
                ForEach(groups) { group in
                    Row(shop: shop, group: group)
                    if group.id != groups.last?.id { Divider() }
                }
            }
        }
    }

    private struct Row: View {
        @Bindable var shop: Shop
        let group: KhaytEngine.PriceGroup

        var body: some View {
            HStack(alignment: .firstTextBaseline, spacing: 10) {
                VStack(alignment: .leading, spacing: 2) {
                    HStack(spacing: 4) {
                        Text(group.material).lineLimit(1)
                        // THE UNIT IS PART OF THE NAME HERE. Two cards reading
                        // "PLA" with different figures would look like a bug;
                        // "PLA /kg" and "PLA /spool" are two honest answers.
                        Text("/" + group.unit)
                            .font(.caption).foregroundStyle(.secondary)
                    }
                    Text(said).font(.caption).foregroundStyle(.secondary).lineLimit(1)
                }
                Spacer(minLength: 8)
                if let move = group.pctChange, abs(move) >= 5 {
                    // Five percent, which is the other app's threshold: a
                    // badge on every one-riyal wobble is a badge nobody reads.
                    Text((move > 0 ? "▲" : "▼") + Money.quantity(abs(move)) + "%")
                        .font(.caption).monospacedDigit()
                        .foregroundStyle(move > 0 ? Khayt.late : Khayt.done)
                }
                if let latest = group.latest {
                    Text(Money.text(latest.price, shop.currency))
                        .font(.callout).monospacedDigit()
                }
            }
            .padding(.vertical, 7)
        }

        /// The second line: how many purchases are behind the figure, who sold
        /// it cheapest, and — when it happened — that the group mixes grams
        /// with kilograms.
        private var said: String {
            var parts = [shop.words.counting(group.count, "mac.purchases_word")]
            if let best = group.best, group.count > 1, !best.supplier.isEmpty {
                parts.append(shop.words.callIt("mac.cheapest_was")
                             + " " + Money.text(best.price, shop.currency)
                             + " · " + best.supplier)
            } else if let latest = group.latest, !latest.supplier.isEmpty {
                parts.append(latest.supplier)
            }
            if group.converted { parts.append(shop.words.callIt("mac.units_converted")) }
            return parts.joined(separator: " · ")
        }
    }
}
